### DISCUSSION TITLE
Support for Multi-Tenant Applications and Direct Database Integration

### DISCUSSION THREAD

**Discussion by raguila8** [tinacms/tinacms#4005] [Ideas]
Hey guys,

Firstly, I want to express my appreciation for Tinacms. I've used it in a few different projects and  It's truly awesome!

I am working on a multi-tenant app that uses the [Vercel's Platform](https://github.com/vercel/platforms) template. My application, as a requirement, is backed by a database and I would like to integrate TinaCMS as a content management system for each of my tenants. However, I think the current setup of TinaCMS may not cater to my specific requirements, especially regarding multi-tenancy and database support. Would implementing something like this be possible with tinacms?

I was thinking about this for a while and referencing the [self-hosted solution and demo](https://github.com/tinacms/tinacms/discussions/3589), but because the tinacms app and client are created at build time, it makes it difficult to support a dynamic architecture. Also, not sure how the `useTina` hook and tinacms datalayer could work with data stored in a database. And then there's the LevelDB ephemeral cache that I think wouldn't be necessary if you're storing the data in a database. 

I would love to hear your thoughts on this. Could something like this be realized with TinaCMS's existing toolset, or are there any plans to support such functionalities in the future?

**Comment by jeffsee55** [comment]
I'm not familiar with Vercel's platform template but it sounds like instead of self-hosting you might want the content to be separate from the code repo, and you could still use Tina Cloud for each client. Each "tenant" would be it's own Github content repo, and the config would change based on that. So you'd have the following Github repos (if you had 2 clients):

```
my-website-template // actual website (eg. NextJS app)
customer-a-content // only stores content
customer-b-content // only stores content
``` 
And in Tina Cloud, you connect the content repos to their own projects
```
customer-a-content -> ClientID: ABC, Token: 123
customer-b-content -> ClientID: DEF, Token: 456
```

And for deployments:
```
customer-a.my-platform.com -> Github repo is `my-website-template` -> Env vars: ABC, 123 
customer-b.my-platform.com -> Github repo is `my-website-template` -> Env vars: DEF, 456
```
So if you're deploying to `customer-a.my-platform.com`, the environment variables would be the `CONTENT_REPO_BRANCH`, `TINA_CLIENT_ID`, and `TINA_TOKEN` that relates to `customer-a`'s repo in Github. 

```
// tina/config.ts
export default defineSchema({
  clientId: process.env.TINA_CLIENT_ID,
  token: process.env.TINA_TOKEN,
  branch: "main" // Note that this value won't be related to the deployment since content is separate
  ...
```
Curious to see if that suites your needs, let us know if that works out. More info on separate content repos:

https://tina.io/guides/tinacms/separate-content-repo/guide/

**Reply by raguila8** [reply]
Storing the data in separate Github content repos might work. I just don't know how manageable this would be and how it would integrate with my own app data and auth. 

This would also require me to create a separate project for each subdomain in order to have different environment variables which would defeat the purpose of having a multi-tenant app I think. As far as I know, you can't conditionally set different env variable values based on the subdomain/domain right? 

So the idea is to only deploy one project where the routing (subdomains and domains) is handled. But since the Tina config lives in this project and not the content repos, I don't know how I could point to the correct repos on a subdomain/domain basis.

**Reply by jeffsee55** [reply]
If you're going to use the same app template for each domain, the Tina schema's would need to be identical right? That would be relatively easy to keep in sync. But to your point about having a project for each subdomain, do you mean a Tina "project" or something else? I would expect each subdomain to be it's own Tina project so users don't mix.

For using different env vars based on domain you wouldn't necessarily need to do that, but I'm not 100% sure on how the platform setup works so sorry if this doesn't make sense:

```ts
// tina/config.ts
const clientIdMap = {
  'customer-a': process.env.TINA_CLIENT_ID_CUSTOMER_A,
  'customer-b': process.env.TINA_CLIENT_ID_CUSTOMER_B
}
const tokenMap = {
  'customer-a': process.env.TINA_TOKEN_CUSTOMER_A,
  'customer-b': process.env.TINA_TOKEN_CUSTOMER_B
}
// Assuming the URLS are `customer-a.vercel.com` https://vercel.com/docs/concepts/projects/environment-variables/system-environment-variables
const customer = new URL(`https://${process.env.VERCEL_URL}`).host.replace(".vercel.com", "")

export default defineSchema({
  clientId: clientIdMap[customer],
  token: tokenMap[customer],
  branch: "main" // Note that this value won't be related to the deployment since content is separate
  ...
```
Would that work?

Here I have hardcoded these "mappings" and you'd need env vars for each customer, but I would expect that these values could just as easily come from a database that your CI process has access to.

**Reply by raguila8** [reply]
Yes, that's correct the tina schema is the same. What I meant about different projects is I would need to create different Vercel projects that each point to the main app repo since I would need different env variables.

The problem with this is that the Tina client and app are built at build time before we know anything about the tenant and their subdomain/domain. That is, when I deploy the app to Vercel, Vercel builds the app with the following build script that runs the Tina cli:

```
// package.json

"scripts": {
    "dev": "tinacms dev -c \"next dev\"",
    "build": "tinacms build && next build",
   //...
  },
```

In more concrete terms, when a request is made to `customer-a.my-platform.com`, the Middleware will inspect the `Host` header, find the value `customer-a.my-platform.com`, and redirect the request to the route that serves customer-a's content. All of this happens after the Tina app has already been built. Does that make sense?

**Reply by jeffsee55** [reply]
Oh yes, I see what you're saying. I assumed you'd be happy with each subdomain having it's own build process, but it sounds like that's not what's needed here. 

I think this _might_ be something we could support, I'd love it if there was a concrete example I could see. Basically - Tina needs the schema at build time, but since the schema will be the same for each site that's not a problem, it's really just that it needs to point to a different client without running a build. We set up the API urls during the build, but that's not a firm requirement so might be something we can fix.

This is pretty similar to the needs we have for implementing branching support, which is still in a somewhat experimental state.

**Reply by raguila8** [reply]
Yeah, maybe the Tina config can have an option to pass in either a callback with a `request` param or a mapper that maps URLs to the correct ids. Something like this:

```ts
// tina/config.ts
export default defineConfig({
  // ...
  cmsCallback: (cms) => {
    cms.clientIdCallback= (request) => {
      // Fetch the correct clientId based on the request url
    },
    cms.tokenCallback= (request) => {
      // Fetch the correct token based on the request url
    }
  }
});
```

I can set up a multi-tenant app example if you'd like, but it essentially will be a clone of this repo: [https://github.com/vercel/platforms](https://github.com/vercel/platforms) with tinacms added to it with posts hooked up to tina with visual editing.
